pkg/volume/csi/csi_plugin.go

```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		// This function prevents Kubelet from posting Ready status until CSINodeInfo
+		// is both installed and initialized
+		err := initializeCSINodeInfo(host)
+		if err != nil {
+			return fmt.Errorf("failed to initialize CSINodeInfo: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func initializeCSINodeInfo(host volume.VolumeHost) error {
+	kvh, ok := host.(volume.KubeletVolumeHost)
+	if !ok {
+		klog.V(4).Infof("Skipping CSINodeInfo initialization, not running on Kubelet")
+		return nil
+	}
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil {
+		// If we're NOT on the Kubelet OR KubeClient DOESNT exist
+		return fmt.Errorf("failed to get KubeClient")
+	}
+
+	kvh.SetKubeletError(fmt.Errorf("CSINodeInfo is not yet available"))
+
+	go func() {
+		defer utilruntime.HandleCrash()
+
+		// Backoff parameters tuned to retry over 140 seconds. Will fail and restart the Kubelet
+		// after max retry steps.
+		crdBackoff := wait.Backoff{
+			Steps:    6,
+			Duration: 15 * time.Millisecond,
+			Factor:   6.0,
+			Jitter:   0.1,
+		}
+		err := wait.ExponentialBackoff(crdBackoff, func() (bool, error) {
+			groupVersion := "csi.storage.k8s.io" + "/" + "v1alpha1"
+			groupResource, err := kubeClient.Discovery().ServerResourcesForGroupVersion(groupVersion)
+			if err != nil {
+				kvh.SetKubeletError(fmt.Errorf("failed to get group resource for group version %v", groupVersion))
+				klog.Errorf("Failed to get groupResource for group version %v: %v", groupVersion, err)
+				return false, nil
+			}
+
+			for _, g := range groupResource.APIResources {
+				if g.Name == "csinodeinfos" {
+					klog.V(4).Infof("Found the csinodeinfos CRD installed on API Server")
+					return true, nil
+				}
+			}
+			return false, nil
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			klog.Fatalf("Failed to find CSINodeInfo on API Server after retrying")
+		}
 
+		kvh.SetKubeletError(fmt.Errorf("CSINodeInfo available. CSINodeInfo not yet initialized"))
+
+		initBackoff := wait.Backoff{
+			Steps:    6,
+			Duration: 15 * time.Millisecond,
+			Factor:   6.0,
+			Jitter:   0.1,
+		}
+		err = wait.ExponentialBackoff(initBackoff, func() (bool, error) {
+			klog.V(4).Infof("Initializing migrated drivers on CSINodeInfo")
+			err := nim.InitializeMigratedDrivers()
+			if err != nil {
+				kvh.SetKubeletError(fmt.Errorf("Failed to initialize CSINodeInfo: %v", err))
+				klog.Errorf("Failed to initialize CSINodeInfo: %v", err)
+				return false, nil
+			}
+
+			// Successfully initialized drivers, allow Kubelet to post Ready
+			kvh.SetKubeletError(nil)
+			return true, nil
+		})
+		if err != nil {
+			// Kill the Kubelet process and allow it to restart to retry initialization
+			klog.Fatalf("Failed to initialize CSINodeInfo after retrying")
```

Comment:
i though that usage of `klog.Fatalf` was discouraged, since if someone vendors such a package their app might exit in a unwanted manner.

also calling exiting from plugin seems quite wrong, in general.